### DISCUSSION TITLE
Nanopore support

### DISCUSSION THREAD

**Discussion by igboyes** [virtool/.github#9] [Ideas]
Document all QC tools that will be required to create a Nanopore sample in Virtool.

_Example_:

# FastQC

## Overview

Will require custom parser for data files as they are not standard formats.

## Graphics
- base-wise quality
- sequence-wise quality
- GC content

## Tabular
- sequence length range
- GC content

## Output Formats
- Custom text-based data file
- HTML

**Comment by peterk87** [comment]
# Nanopore sequencing data

**WIP:** I'll update this when I get the chance next week. 

## Background

FAST5 is an [HDF5] file format output by Oxford Nanopore Technologies (ONT) sequencing instruments, which primarily stores read raw current signal data when a nucleotide molecule passes through a pore. A typical FAST5 is around 10-20X the size of the equivalent basecalled FASTQ. The FAST5 format also contains a lot of metadata about the sequencing run and each read. FAST5 can also be used to store the basecalled sequence data as well as "events" identified by certain software.

When basecalling of Nanopore sequencing data is performed with [Guppy], some of the metadata present in the FAST5 is written to the read headers in the FASTQ file containing basecalled sequencing data.

> There are alternative basecallers to Guppy, but it's generally recommended to use Guppy. See [Performance of neural network basecalling tools for Oxford Nanopore sequencing](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1727-y) and [the August 2019 update](https://github.com/rrwick/August-2019-consensus-accuracy-update)

## Output Files

Below is an example of the file and directory structure of a Nanopore sequencing run with basecalling and barcode demultiplexing enabled (like the GridION):

```
.
├── drift_correction_${flowcell_id}_${acquisition_run_id[:9]}.csv
├── duty_time_${flowcell_id}_${acquisition_run_id[:9]}.csv
├── fast5_fail
│   ├── barcode17
│   │   └── ${flowcell_id}_fail_barcode17_${acquisition_run_id[:9]}_0.fast5
│   ├── barcode21
│   │   └── ${flowcell_id}_fail_barcode21_${acquisition_run_id[:9]}_0.fast5
│   └── unclassified
│       └── ${flowcell_id}_fail_unclassified_${acquisition_run_id[:9]}_0.fast5
├── fast5_pass
│   ├── barcode15
│   │   └── ${flowcell_id}_pass_barcode15_${acquisition_run_id[:9]}_0.fast5
│   ├── barcode16
│   │   └── ${flowcell_id}_pass_barcode16_${acquisition_run_id[:9]}_0.fast5
│   ├── barcode19
│   │   └── ${flowcell_id}_pass_barcode19_${acquisition_run_id[:9]}_0.fast5
│   ├── barcode20
│   │   └── ${flowcell_id}_pass_barcode20_${acquisition_run_id[:9]}_0.fast5
│   ├── barcode21
│   │   └── ${flowcell_id}_pass_barcode21_${acquisition_run_id[:9]}_0.fast5
│   └── unclassified
│       └── ${flowcell_id}_pass_unclassified_${acquisition_run_id[:9]}_0.fast5
├── fastq_fail
│   ├── barcode17
│   │   └── ${flowcell_id}_fail_barcode17_${acquisition_run_id[:9]}_0.fastq
│   ├── barcode21
│   │   └── ${flowcell_id}_fail_barcode21_${acquisition_run_id[:9]}_0.fastq
│   └── unclassified
│       └── ${flowcell_id}_fail_unclassified_${acquisition_run_id[:9]}_0.fastq
├── fastq_pass
│   ├── barcode15
│   │   └── ${flowcell_id}_pass_barcode15_${acquisition_run_id[:9]}_0.fastq
│   ├── barcode16
│   │   └── ${flowcell_id}_pass_barcode16_${acquisition_run_id[:9]}_0.fastq
│   ├── barcode19
│   │   └── ${flowcell_id}_pass_barcode19_${acquisition_run_id[:9]}_0.fastq
│   ├── barcode20
│   │   └── ${flowcell_id}_pass_barcode20_${acquisition_run_id[:9]}_0.fastq
│   ├── barcode21
│   │   └── ${flowcell_id}_pass_barcode21_${acquisition_run_id[:9]}_0.fastq
│   └── unclassified
│       └── ${flowcell_id}_pass_unclassified_${acquisition_run_id[:9]}_0.fastq
├── final_summary_${flowcell_id}_${acquisition_run_id[:9]}.txt
├── mux_scan_data_${flowcell_id}_${acquisition_run_id[:9]}.csv
├── report_${flowcell_id}_20201009_2115_cff09992.md
├── report_${flowcell_id}_20201009_2115_cff09992.pdf
├── sequencing_summary_${flowcell_id}_${acquisition_run_id[:9]}.txt
└── throughput_${flowcell_id}_${acquisition_run_id[:9]}.csv
```

### final_summary.txt

The presence of a `final_summary*.txt` file can be used to indicate the completion of a Nanopore sequencing run. It also contains a lot of useful information that could be parsed.

Example `final_summary*.txt` for a GridION:

```
instrument=${instrument_id}
position=X1
flow_cell_id=${flowcell_id}
sample_id=${run_id}
protocol_group_id=${run_id}
protocol=sequencing/sequencing_MIN106_DNA:FLO-FLG001:SQK-LSK109
protocol_run_id=${uuid}
acquisition_run_id=${hex_id_40_characters}
started=${isoformat_datetime}
acquisition_stopped=${isoformat_datetime}
processing_stopped=${isoformat_datetime}
basecalling_enabled=1
sequencing_summary_file=sequencing_summary_${flowcell_id}_${acquisition_run_id[:9]}.txt
fast5_files_in_final_dest=9
fast5_files_in_fallback=0
fastq_files_in_final_dest=9
fastq_files_in_fallback=0
```

### sequencing_summary.txt

A `sequencing_summary*.txt` is generated by both MinKNOW UI frontend for Nanopore sequencing and the Guppy basecaller is a tab-delimited table which can contain:

- sequencing information for each read (when it went through the pore, how long it took to go through the pore, etc)
- whether it passes Q-score filtering (if enabled)
- Guppy basecalling information
- Guppy barcode assignment info


Field | Value
----- | -----
filename_fastq | filename.fastq
filename_fast5 | filename.fast5
read_id | READ_UUID
run_id | RUN_ID
channel | 5
mux | 1
start_time | 100.59325
duration | 0.62475
num_events | 499
passes_filtering | FALSE
template_start | 100.64825
num_events_template | 455
template_duration | 0.56975
sequence_length_template | 215
mean_qscore_template | 4.104715
strand_score_template | 0
median_template | 141.270203
mad_template | 9.213274
pore_type | not_set
experiment_id | EXPERIMENT_ID
sample_id | SAMPLE_ID
end_reason | signal_positive
barcode_arrangement | unclassified
barcode_full_arrangement | NB18_var1
barcode_kit | NB
barcode_variant | var1
barcode_score | 21.82222
barcode_front_id | NB18_FWD
barcode_front_score | 0.307693
barcode_front_refseq | AGGTTAACCAAACCCAACAACCTAGATAGGCCAGCACCT
barcode_front_foundseq | ATTTAAAAAGCATAAGTATCCGTTATACACACT
barcode_front_foundseq_length | 33
barcode_front_begin_index | 72
barcode_rear_id | NB18_REV
barcode_rear_score | 21.82222
barcode_rear_refseq | GGTGCTGGCCTATCTAGGTTGTTGGGTTTGGTTAACCTTAGCAAT
barcode_rear_foundseq | TGCTACCATTCAGCACGTTTTCGTTCTTCTAACCTATATGCT
barcode_rear_foundseq_length | 42
barcode_rear_end_index | 5

> The `sequencing_summary*.txt` files can be large uncompressed (>100MB).

### Guppy Basecalled FASTQ

Some of the information contained within the `sequencing_summary` and `final_summary` tables can be found in the Guppy basecalled FASTQ read headers, e.g.:

```
@${read_uuid} runid=${acquisition_run_id} sampleid=${sample_id} read=${read_number} ch=${channel_number} start_time=${utc_datetime} barcode=barcode01
```

## User Upload Requirements

If users are minimally required to upload their basecalled (and demultiplexed) FASTQ files (with some metadata in the FASTQ read headers), then it wouldn't be too difficult to calculate stats and create plots from the info embedded within the read headers, e.g. read start time by read length with colour fill by quality score.

[NanoPlot](https://github.com/wdecoster/NanoPlot) and [pycoQC](https://github.com/a-slide/pycoQC) analysis and figure generation with a `sequencing_summary*.txt` accompanying the sample FASTQs. However, these tools don't seem to output easily parseable data formats, so it may be easier to roll our own.

## Example Python code to parse Guppy basecalled FASTQ read info

```python
import pandas as pd
from Bio import SeqIO
from Bio import SeqRecord

def parse_nanopore_read_stats(fq_record: SeqRecord) -> dict: 
    # mean Q-score, read length, date/time, seq start time
    # tokenize/split fq_record.description on '=' into dict

# read stats records into Pandas DataFrame
df_read_stats = pd.DataFrame(parse_nanopore_read_stats(fq_record) for fq_record in SeqIO.parse(fastq, 'fastq'))
# stats about length; quartiles, mean, stdev
df_read_stats.read_length.describe().to_json(orient='records')
# JSON string of records for viz
df_read_stats[['read_length, 'start_time', 'q_score']].to_json(orient='records')
```



[HDF5]: https://github.com/nanoporetech/ont_fast5_api
[FAST5]: https://www.hdfgroup.org/solutions/hdf5
[Guppy]: https://community.nanoporetech.com/
